druid/src/text/editable_text.rs

```diff
@@ -156,6 +162,32 @@ impl EditableText for String {
     fn from_str(s: &str) -> Self {
         s.to_string()
     }
+
+    fn line_beginning(&self, from: usize) -> Option<usize> {
+        let mut offset = from;
+
+        for grapheme in self.get(0..from)?.graphemes(true).rev() {
+            if grapheme.chars().next()? == '\n' {
+                return Some(offset);
+            }
+            offset -= grapheme.len();
+        }
+
+        None
+    }
+
+    fn line_ending(&self, from: usize) -> Option<usize> {
+        let mut offset = from;
+
+        for grapheme in self.get(from..)?.graphemes(true) {
+            if grapheme.chars().next()? == '\n' {
+                return Some(offset);
+            }
+            offset += grapheme.len();
+        }
+
+        Some(self.len())
```

Comment:
Same as above, except that we can just return `self.len()` if we don't find a hit.